app/workers/messaging.php

```diff
@@ -20,59 +30,180 @@
+        foreach ($users as $user) {
+            $recipients = \array_merge($recipients, $user->getAttribute('targets'));
         }
 
-        $message = new SMS(
-            to: [$this->args['recipient']],
-            content: $this->args['message'],
-            from: $this->from,
-        );
+        $targets = $this->dbForProject->find('targets', [Query::equal('$id', $recipientsId)]);
+        $recipients = \array_merge($recipients, $targets);
 
-        try {
-            $this->sms->send($message);
-        } catch (\Exception $error) {
-            throw new Exception('Error sending message: ' . $error->getMessage(), 500);
+        $identifiers = \array_map(function (Document $recipient) {
+            return $recipient->getAttribute('identifier');
+        }, $recipients);
+
+        $maxBatchSize = $provider->getMaxMessagesPerRequest();
+        $batches = \array_chunk($identifiers, $maxBatchSize);
+        $deliveredTo = 0;
+
+        foreach ($batches as $batch) {
+            $messageRecord->setAttribute('to', $batch);
+            $message = match ($providerRecord->getAttribute('type')) {
+                'sms' => $this->buildSMSMessage($messageRecord, $providerRecord),
+                'push' => $this->buildPushMessage($messageRecord),
+                'email' => $this->buildEmailMessage($messageRecord, $providerRecord),
+                default => null
+            };
+            try {
+                $provider->send($message);
+                $deliveredTo += \count($batch);
+            } catch (Exception $e) {
+                $deliveryErrors = $messageRecord->getAttribute('deliveryErrors');
+                foreach ($batch as $identifier) {
+                    $deliveryErrors[] = 'Failed to send message to target' . $identifier . ': ' . $e->getMessage();
+                }
+                $messageRecord->setAttribute('deliveryErrors', $deliveryErrors);
+            }
         }
+
+        if (\count($messageRecord->getAttribute('deliveryErrors')) > 0) {
+            $messageRecord->setAttribute('status', 'failed');
+        } else {
+            $messageRecord->setAttribute('status', 'sent');
+        }
+        $messageRecord->setAttribute('to', $recipientsId);
+        $messageRecord->setAttribute('deliveredTo', $deliveredTo);
+        $messageRecord->setAttribute('deliveryTime', DateTime::now());
+
+        $this->dbForProject->updateDocument('messages', $messageRecord->getId(), $messageRecord);
     }
 
     public function shutdown(): void
     {
     }
+
+    private function buildEmailMessage($message, $provider): Email
+    {
+        $from = $provider['options']['from'];
+        $to = $message['to'];
+        $subject = $message['data']['subject'];
+        $content = $message['data']['content'];
+        $html = $message['data']['html'];
+        return  new Email(to: $to, subject: $subject, content: $content, from: $from, html: $html);
```

Comment:
Let's line break and remove redundant parameter names and the extra space

Reply:
done this